org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BuildNotifier.java

```diff
@@ -76,8 +89,35 @@ public void begin() {
  * Check whether the build has been canceled.
  */
 public void checkCancel() {
-	if (this.monitor != null && this.monitor.isCanceled())
-		throw new OperationCanceledException();
+	if (this.monitor != null && this.monitor.isCanceled()) {
+		// User requested cancel.
+		if (this.buildKind == IncrementalProjectBuilder.AUTO_BUILD) {
```

Comment:
I'm not sure we should be delaying user-induced cancellation: the user is boss! Anyway...the chance that a user can click cancel on an autobuild job before one second has elapsed seems small to me.

Reply:
thats an excellent question.  Think of my workspace with ~500 projects where the autobuild can take several minutes. the user wants to cancel the autobuild because he does not want to wait minutes, but he does not care about a second. However if he presses cancel he has little luck to cancel autobuild between 2 projects. Higher chance he cancels just while a single project is building. That would result in a complete unbuild project which could take minutes to full-build on next autobuild. while when he just waited 1 second to get a good moment between projects then all autobuilds can resume incremental in the next round.

Reply:
@jukzi Why don't we cancel at the first project state save operation to avoid a full build ? Or if you are trying to avoid a full build because we cancel before a delta build finishes, may be we can check for project state write with the build kind or cancel if the time elapse, which ever occurs first ?

Reply:
@gayanper that sound like you got a better idea - however my jdt knowledge is to limited to fully understand what that meand or how to implement your suggestion. i simply tried to extend the obvious existing cancel checks.
Could you please provide a alternative solution if my change is too bad?

Reply:
@jukzi your change is good, i just want to make sure that we will not cancel in middle of delta changes which could result in full builds. I’m not sure the timeout we have right now in this patch is sufficient for all cases. There is a method that does update the build state at the end of delta builds. May be you can check that to see if that can be use to cancel without waiting and let the current logic handle all other scenarios.

i also came with these suggestions when i checked the code after reviewing this patch. So take my suggestion with a pinch of salt 😉

Reply:
Do you mean org.eclipse.jdt.internal.core.builder.JavaBuilder.buildDeltas(SimpleLookupTable)? it does not sound as an improvement to me to add there another cancel check as it is not within any loop. between projects the BuildManager will cancel immediatly anyway. org.eclipse.core.internal.events.BuildManager.checkCanceled(int, IProgressMonitor)